google/cloud/storage/benchmarks/benchmark_utils.cc

```diff
@@ -118,6 +119,78 @@ std::chrono::seconds ParseDuration(std::string const& val) {
   return std::chrono::seconds(s);
 }
 
+bool ParseBoolean(std::string const& val, bool default_value) {
+  if (val.empty()) {
+    return default_value;
+  }
+  auto lower = val;
+  std::transform(lower.begin(), lower.end(), lower.begin(),
+                 [](char x) { return static_cast<char>(std::tolower(x)); });
+  if (lower == "true") {
+    return true;
+  } else if (lower == "false") {
+    return false;
+  }
+  throw std::runtime_error("Cannot parse " + val + " as a boolean");
+}
+
+std::string Basename(std::string const& path) {
+  // With C++17 we would use std::filesytem::path, until then do the poor's
+  // person version.
+#if _WIN32
+  return path.substr(path.find_last_of('\\') + 1);
+#else
+  return path.substr(path.find_last_of('/') + 1);
+#endif  // _WIN32
+}
+
+std::string BuildUsage(std::vector<OptionDescriptor> const& desc,
+                       std::string const& command_path) {
+  std::ostringstream os;
+  os << "Usage: " << Basename(command_path) << " [options] <region>\n";
+  for (auto const& d : desc) {
+    os << "    " << d.option << ": " << d.help << "\n";
+  }
+  return std::move(os).str();
+}
+
+std::vector<std::string> OptionsParse(std::vector<OptionDescriptor> const& desc,
+                                      std::vector<std::string> argv) {
+  if (argv.empty()) {
+    return argv;
+  }
+  std::string const usage = BuildUsage(desc, argv[0]);
+
+  auto next_arg = argv.begin() + 1;
+  while (next_arg != argv.end()) {
+    std::string const& argument = *next_arg;
+
+    // Try to match argv[next_arg] against the options in `desc`
+    bool matched = false;
+    for (auto const& d : desc) {
+      if (argument.rfind(d.option, 0) != 0) {
+        // Not a match, keep searching
+        continue;
+      }
+      std::string val;
+      if (d.option.size() + 1 < argument.size()) {
+        val = argument.substr(d.option.size() + 1);
+      }
+      d.parser(val);
+      // This is a match, consume the argument and stop the search.
+      matched = true;
+      argv.erase(next_arg);
+      next_arg = argv.begin() + 1;
```

Comment:
`next_arg = argv.erase(next_arg);`?

Although I would be tempted to move that below into a single `next_arg` update.

`next_arg = matched ? argv.erase(next_arg) : next_arg + 1;`

Reply:
I am not sure I follow. I made the first change, but `next_arg = next_arg + 1` should only happen if the `for (d : desc)` loop terminates without finding a match, so it needs to be outside the for loop?  I just probably misread what you meant.

Reply:
> should only happen if the for (d : desc) loop terminates without finding a match

Right ... that's why the suggestion tests `matched`.  Unless I'm not following of course.

Reply:
The whole suggestion is supposed to be outside the loop, yes. That's what I meant by "move that below", but sorry for not being clear.

Reply:
I think I got it this time.